modernize porkbun-ddns

con-f-use opened the pull request:
Modernizes the Python DDNS client:

* Use argparse for CLI argument handling (better help text, validation and error messages)
* Accept config file from stdin
* Use modern Python naming conventions
* Use ipaddress for validation
* Default value for the base url of the API
* Hashbang and magic `__main__` for use as executable script
* Closes #2 
* Closes #3

con-f-use committed:
modernize porkbun-ddns

Modernizes the Python DDNS client:

* Use argparse for CLI argument handling (better help text, validation and error messages)
* Accept config file from stdin
* Use modern Python naming conventions
* Use ipaddress for validation
* Default value for the base url of the API
* Closes #2 
* Closes #3

con-f-use commented:
@rossporkbun @eddie-porkbun please review. I tried to keep the CLI as much as possible - at some point, in a later PR, I'd like to add braking changes if that is okay with you:

Namely, a default config file location in `~/.config/porkbun/ddns.conf` that can be overwritten via environment variable and `-c, --config` option as well as the possibility to save domain and subdomain there and overwrite everything with their own command line options. Also: logging, verbosity option and maybe a `setup.py` and `flake.nix` for packaging and easy distribution. But only if you're interested.

benneti commented:
If I try to use this with -i IPV6 adress, it set's the A record, I guess it should properly format the args when the ip is given.

con-f-use commented:
@benneti are you sure, because there is: https://github.com/con-f-use/porkbun-dynamic-dns-python/blob/patch-1/porkbun-ddns.py#L52 in the code.

benneti commented:
Thanks for the quick reply.
Yeah I saw this, too and was surprised that it nonetheless chose A.
I can double check later whats args contains if I use -i.

On 12 February 2022 11:02:58 CET, con-f-use ***@***.***> wrote:
***@***.*** are you use, because there is: https://github.com/con-f-use/porkbun-dynamic-dns-python/blob/patch-1/porkbun-ddns.py#L52 in the code.
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/porkbundomains/porkbun-dynamic-dns-python/pull/6#issuecomment-1037101563
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***>
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

benneti commented:
ok sorry it works as intended now, I guess I mixed up the lines.
The actual problem is that if I set the AAAA record it also deletes the A record, but I would like to have both set at the same time.
Basically I think
```python
def delete_record(args):
    type_ = "A" if args.public_ip.version == 4 else "AAAA"
    for i in get_records(args)["records"]:
        if i["name"] == args.fqdn and i["type"] in [type_, "ALIAS", "CNAME"]:
            print("Deleting existing {}-Record: {}".format(i["type"], i))
            api(args, "/dns/delete/" + args.domain + "/" + i["id"])
```
would be better suited

con-f-use commented on porkbun-ddns.py:
```diff
@@ -1,53 +1,94 @@
+    return all_records
+
+
+def delete_record(args):
+    for i in get_records(args)["records"]:
+        if i["name"] == args.fqdn and i["type"] in ["A", "AAAA", "ALIAS", "CNAME"]:
+            print("Deleting existing {}-Record: {}".format(i["type"], i))
+            api(args, "/dns/delete/" + args.domain + "/" + i["id"])
```
```suggestion
def delete_record(args):
    type_ = "A" if args.public_ip.version == 4 else "AAAA"
    for i in get_records(args)["records"]:
        if i["name"] == args.fqdn and i["type"] in [type_, "ALIAS", "CNAME"]:
            print("Deleting existing {}-Record: {}".format(i["type"], i))
            api(args, "/dns/delete/" + args.domain + "/" + i["id"])
```

con-f-use commented:
@benneti
Did you know, you can make code suggestions in comments:

![image](https://user-images.githubusercontent.com/11145016/153753680-61a9fd84-7b34-45b6-8145-e663f0b9fe33.png)

You just need to be in !["Files Changed"](https://user-images.githubusercontent.com/11145016/153753989-1b855f66-a062-4f1e-82e8-04b649c70833.png) and then click on a staring line ![green plus icon](https://user-images.githubusercontent.com/11145016/153754048-982d43b2-5833-4608-81f5-46ff0e3a0430.png) and drag to the last line your change is about.

con-f-use committed:
only delete matching type (by benneti)

suggested in https://github.com/porkbundomains/porkbun-dynamic-dns-python/pull/6#issuecomment-1037362375 by @benneti

benneti commented:
I did not, but I think I could also have done a review to achieve the same.
Thanks for putting it in the PR anyway!

rossporkbun commented:
@con-f-use sorry for the delay in this response. Thank you so much for doing all this work, and I bow to your Python knowledge, it bests mine. 

I'm guessing this can't be made Python 2 compliant, which is a necessary requirement for the program. If you can, I'd be glad to consider the pull request. Otherwise, if you want to release the client on your own, I will happily link to your repo in our official documentation. That way, you can maintain it yourself rather than relying on me to approve your pull requests, as I am admittedly slow to take action.

con-f-use commented:
I think it can be lightly tweaked to work with Python 2 AND 3. I'll see what I can do.

That being said, a word of advice: Python 2 has been end of life for close to ~two~ three years non and does not get any updates, security or otherwise. You should really switch any remaining infrastructure, because soon you'll face increasing problems with even installing it, since a lot of tools (pip, setuptools and the python package index) will or have already phased out support. I saw those problems daily in the company I work for.

con-f-use committed:
Merge branch 'main' into patch-1

con-f-use commented:
It does work with python2.7 as long as [`py2-ipaddress`](https://pypi.org/project/py2-ipaddress/) and [`requests`](https://pypi.org/project/requests/2.27.1/#description) are installed in the right versions.

mietzen commented:
@rossporkbun any Update on this? This would be highly beneficial for all users with IPv6 dual-stack lite connection which getting more common.

dramborleg commented:
> @rossporkbun any Update on this? This would be highly beneficial for all users with IPv6 dual-stack lite connection which getting more common.

Given the low activity I'd be surprised to see a response in the near future, but just in case it happens to help you I figured I would mention that ddclient recently added porkbun support in https://github.com/ddclient/ddclient/issues/437. Not sure if it's in an official ddclient release yet.

eddie-porkbun commented:
> @rossporkbun any Update on this? This would be highly beneficial for all users with IPv6 dual-stack lite connection which getting more common.

Ross actually has left Porkbun some time ago for new adventures. Ultimately this was meant to be an example of how to use our API for DDNS, provided as is, rather than an officially supported client. Using ddclient is probably a better way to go and I'll be sure to test it out myself soon.

mietzen commented:
@eddie-porkbun 
I think it would be a good Idea to update the ReadMe, reference ddclient and archive this repo.
Since this repo is still one of the first hits when you Google ddns and porkbun.
